### ISSUE TITLE
Discussion : Hashing algorithm other than SHA-1

### ISSUE THREAD

**Issue by OlivierJaquemet** [samdjstevens/java-totp#27]
Were you able to successfully use hashing algorithm other than SHA-1 with any application ? 
I could not...

I have tested the following application and I could not make it work : 

On Android : 
-  [andOTP - Android OTP Authenticator](https://play.google.com/store/apps/details?id=org.shadowice.flocke.andotp)  0.6.3.1-play
- [FreeOTP Authenticator](https://play.google.com/store/apps/details?id=org.fedorahosted.freeotp) 1.5
- [Google Authenticator ](https://play.google.com/store/apps/details?id=com.google.android.apps.authenticator2) 5.0
(Though its documentation clearly states it is not supported https://github.com/google/google-authenticator/wiki/Key-Uri-Format#algorithm) 
- [Microsoft Authenticator](https://play.google.com/store/apps/details?id=com.azure.authenticator) 6.2002.0898

On Windows : 
- [KeePassXC](https://keepassxc.org/) 2.5.3-Win64-portable
it is supposed to be supported since https://github.com/keepassxreboot/keepassxc/pull/2972, there is a configuration for it but the do not match java-totp generation
![image](https://user-images.githubusercontent.com/618861/75361778-1e12d980-58b8-11ea-9004-1b1c0045da60.png)

Do you known any application supporting hashing algorithm other than SHA-1 ?

**Comment by Stexxen** [comment]
I think their are 2 problems here. 

App makers are not clear on whether they support SHA-256/512 and I personally I think the [getDigitsFromHash](https://github.com/samdjstevens/java-totp/blob/master/src/main/java/dev/samstevens/totp/code/DefaultCodeGenerator.java#L70) function in the DefaultCodeGenerator has an error for SHA-256/512.

I used [oathtool](https://www.nongnu.org/oath-toolkit/) to verify the results, and I also couldn't make it work. (SHA-1 matched, SHA-256/512 did not)
I eventually replaced it with the below  which I took from the similiar [aerogear function](https://github.com/aerogear-attic/aerogear-otp-java/blob/3e86f7ae7845af63708c7394d9eba2ccb851e0ec/src/main/java/org/jboss/aerogear/security/otp/Totp.java#L132), this then matched the oathttool for all hash types.

```
  private String getDigitsFromHash(byte[] hash) {
    // put selected bytes into result int
    int offset = hash[hash.length - 1] & 0xf;

    int binary = ((hash[offset] & 0x7f) << 24) |
                         ((hash[offset + 1] & 0xff) << 16) |
                         ((hash[offset + 2] & 0xff) << 8) |
                         (hash[offset + 3] & 0xff);

    return String.format("%0" + digits + "d", binary % (int) Math.pow(10, digits));
  }

```

Back to your first question, apps that I have verified working with at SHA-256 **_after_** this change are

Android 

- Oracle Mobile Authenticator 
- Auth0 Guardian 
- Sophos Authenticator 
- FreeOTP Authenticator 

iOS 

- Oracle Mobile Authenticator 
- IBM Verify 
- Sophos Authenticator 
- FreeOTP Authenticator

**Comment by samdjstevens** [comment]
Thanks for raising the issue @OlivierJaquemet and for your investigation/fix @Stexxen !

Will try and get some unit tests in place for all the different hash types and make a fix if needed as soon as I can.

**Comment by Stexxen** [comment]
I have this small test for SHA-256 if its any help?
oathtool result
```
% oathtool --totp=sha256 --now="2019-09-04 21:12:16 UTC" --digits=6 --base32 W3C5B3WKR4AUKFVWYU2WNMYB756OAKWY
272978
stexxen@MacBookPro16 ~ % oathtool --totp=sha256 --now="2019-09-04 21:12:16 UTC" --digits=8 --base32 W3C5B3WKR4AUKFVWYU2WNMYB756OAKWY
76272978
```

and the Test Case
```
  @Test
  public void testSHA256Result() {
    // 2019-09-04 21:12:16 UTC
    int time = 1567631536;
    long currentBucket = Math.floorDiv(time, 30);

    CodeGenerator g = new DefaultCodeGenerator(HashingAlgorithm.SHA256);
    String code = g.generate("W3C5B3WKR4AUKFVWYU2WNMYB756OAKWY", currentBucket);
    assertEquals("272978", code);

    g = new DefaultCodeGenerator(HashingAlgorithm.SHA256, 8);
    code = g.generate("W3C5B3WKR4AUKFVWYU2WNMYB756OAKWY", currentBucket);
    assertEquals("76272978", code);
}
```

**Comment by OlivierJaquemet** [comment]
Awesome work @Stexxen, thank you !

Thank you @samdjstevens for taking the time to maintain your project. If I can help, do not hesitate to ask.

**Comment by samdjstevens** [comment]
I've just pushed a fix for this to master with some unit tests that should hopefuly ensure it's working for all hashing algorithms. Please do have a look over if you can, and if all good I will publish v1.7 shortly.

Thanks!

**Comment by Stexxen** [comment]
Ahhh that's the [problem](https://github.com/samdjstevens/java-totp/commit/51a4d84b79164fd92d87accd38c5c677c25f63c2#diff-114187d5ef94fca3e7eb4a9d2d44e517L71) was it. My binary maths was just not up to the challenge of working out what was wrong...

Thanks for fixing 👍

**Comment by samdjstevens** [comment]
Just released v1.6.1 to maven central with this fix

**Comment by OlivierJaquemet** [comment]
Great work. Version 1.6.1 tested and approved. :)
Other hashing algorithm are working perfectly fine with FreeOTP, andOTP and KeePassXC 👍🏻
Thanks again.
